infrasim/dae.py

```diff
@@ -0,0 +1,628 @@
+        """
+        return {"phy": self.phy,
+                "num": self.num,
+                "atta_type": self.atta_type,
+                "atta_phy": self.atta_phy,
+                "atta_wwn": self.atta_wwn,
+                "atta_dev_name": self.atta_dev_name,
+                "atta_slot_id": self.atta_slot}
+
+    def update_link(self, atta_type, atta_phy, atta_wwn, atta_dev_name):
+        """
+            Overwrite link's information with given attributes
+        """
+        self.atta_type = atta_type
+        self.atta_phy = atta_phy
+        self.atta_wwn = atta_wwn
+        self.atta_dev_name = atta_dev_name
+
+
+class _Drv(object):
+    def __init__(self, cfg, index):
+        self.__link = []
+        self.__cfg = cfg.copy()
+        self.__cfg['slot_number'] = index
+        self.__cfg['file'] = cfg['file'].format(index)
+        # each drive occupis 4 wwn. base wwn, 2 port wwns, target wwn.
+        self.__cfg['wwn'] = cfg['wwn'] + 4 * index
+        self.__cfg['serial'] = _SNGenerator.getSN()
+
+    def connect(self, link):
+        """
+            append link to the drive's link list
+        """
+        self.__link.append(link)
+
+    def get_wwn(self, port_index):
+        """
+            get port wwn in terms of port index.
+        """
+        return self.__cfg['wwn'] + port_index + 1
+
+    def get_dev_name(self):
+        """
+            get drive's device name
+        """
+        # get base wwn.
+        return self.__cfg['wwn']
+
+    def print_link(self):
+        for l in self.__link:
+            print(l)
+
+    def __str__(self, *args, **kwargs):
+        return object.__str__(self.__cfg)
+
+    def get_port_cfg(self, port_index):
+        """
+            return a map copy for specific port index. fill device info attached
+            to the drive's port
+        """
+        _cfg = self.__cfg.copy()
+        _cfg['port_wwn'] = self.get_wwn(port_index)
+        for link in self.__link:
+            if link.phy == port_index:
+                _cfg['atta_wwn'] = link.atta_wwn
+                _cfg['atta_phy_id'] = link.atta_phy
+                break
+        return _cfg
+
+
+class _Expander(object):
+    def __init__(self, cfg, dae_type):
+        self.__cfg = cfg
+        self.__link = []
+        self.name = cfg['name']
+        self.phy_count = cfg['phy_count']
+        # attributes of ses
+        self.__ses_attributes = {}
+        self.__ses_attributes['channel'] = 0
+        self.__ses_attributes['dae_type'] = dae_type
+        self.__ses_attributes['ep_atta_sas_addr'] = 0  # reset it later
+        self.__ses_attributes['lun'] = 0
+        self.__ses_attributes['physical_port'] = 0  # reset it later
+        self.__ses_attributes['pp_atta_sas_addr'] = 0  # reset it later
+        self.__ses_attributes['scsi-id'] = 0  # reset it later
+        self.__ses_attributes['side'] = cfg['side']
+        self.__ses_attributes['wwn'] = cfg['wwn'] - 1  # ses_wwn = exp_wwn - 1
```

Comment:
❓ an open discussion: if it's better to encapsulate a `ses` class

Reply:
same concern here during the implement.
We treat it as a data set of attribute since we think it is a result of process, instead of a object take part in traversal algorithm.